### PR TITLE
feat(kernels): implement polygon triangulation routines

### PR DESCRIPTION
**Pull request by imrn99**
Implement two polygon triangulation methods:
- ear clipping, which works on all polygons with no hole
- fanning, which only works on a subset of polygons; there is a generic implementation of this method and one dedicated to convex polygons

## Scope

- [x] Code

## Type of change

- [x] New feature(s)

### CODE REVIEW

File: honeycomb-kernels/src/triangulation/ear_clipping.rs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+) {
+    let mut n = Orbit2::new(cmap, OrbitPolicy::Custom(&[1]), face_id as DartIdentifier).count();
+
+    // early rets
+    if n <= 3 {
+        println!("I: face {face_id} is an {n}-gon -- skipping triangulation");
+        return;
+    }
+    if (n - 3) * 2 != new_darts.len() {
+        println!("W: not enough pre-allocated darts to triangulate face {face_id} -- skipping triangulation");
```
Comment (cedricchevalier19) [inline honeycomb-kernels/src/triangulation/ear_clipping.rs:57]: Is it always "not enough"?

File: honeycomb-kernels/src/triangulation/fan.rs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    let n = darts.len();
+
+    // early rets
+    if n <= 3 {
+        println!("I: face {face_id} is an {n}-gon -- skipping triangulation");
+        return;
+    }
+    if (n - 3) * 2 != new_darts.len() {
+        println!("W: not enough pre-allocated darts to triangulate face {face_id} -- skipping triangulation");
+        return;
```
Comment (cedricchevalier19) [inline honeycomb-kernels/src/triangulation/fan.rs:49]: Same here.

File: honeycomb-kernels/src/triangulation/fan.rs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    let vertices: Vec<_> = darts
+        .iter()
+        .map(|dart_id| {
+            cmap.vertex(cmap.vertex_id(*dart_id))
+                .expect("E: found a topological vertex with no associated coordinates")
+        })
+        .collect();
```
Comment (cedricchevalier19) [inline honeycomb-kernels/src/triangulation/fan.rs:58]: Duplicated code with `ear_clipping.rs`. Does a function make sense?

File: honeycomb-kernels/src/triangulation/fan.rs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+    // early rets
+    if n <= 3 {
+        println!("I: face {face_id} is an {n}-gon -- skipping triangulation");
+        return;
+    }
+    if (n - 3) * 2 != new_darts.len() {
+        println!("W: not enough pre-allocated darts to triangulate face {face_id} -- skipping triangulation");
+        return;
+    }
```
Comment (cedricchevalier19) [inline honeycomb-kernels/src/triangulation/fan.rs:153]: Same here.

**Review by cedricchevalier19 — COMMENTED** [review]
Nothing is blocking the PR, but I think you can improve the code before merging.

**Comment by imrn99** [comment]
@cedricchevalier19 I added three functions:
- `check_requirements`: this takes care of some of the duplicated code (number of vertex per face, number of pre-allocated darts)
- `fetch_face_vertices`: if all vertices of a face are defined, return a collection with those, else raise en error
- `crossp_from_verts`: this replace the macro; it should be inlined anyway

I also added the `TriangulateError` to describe these issues. The routines now return a `Result<(), TriangulateError>` that can be processed to shut down the program or attempt triangulation with another method.

**Review by cedricchevalier19 — APPROVED** [review]
LGTM

**Pull request merged**